### ISSUE TITLE
Modsecurity POST request with binary file and header "application/x-www-form-urlencoded"

### ISSUE THREAD

**Issue by regazzoj** [SpiderLabs/ModSecurity#2695]
**Describe the bug**

Well I don't know if it is a bug, but this behavior is really strange.
One of my colleague tried to make a POST request with a binary file (which is just a 1MB JSON file), the processing last about ~3 minutes before returning a 403 response. 

**Logs and dumps**

Output of:
 1. DebugLogs (level 9) -> Too big for copy-paste... And I can't figure out what to use
 2. AuditLogs -> Can't access
 3. Error logs -> `2022/03/11 14:00:48 [notice] 1#1: ModSecurity-nginx v1.0.2 (rules loaded inline/local/remote: 0/1828/0)
2022/03/11 14:01:30 [error] 19#19: *1 ModSecurity: Access denied with code 403 (phase 2). Matched "Operator Ge' with parameter 5' against variable TX:ANOMALY_SCORE' (Value: 10' ) [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-949-BLOCKING-EVALUATION.conf"] [line "80"] [id "949110"] [rev ""] [msg "Inbound Anomaly Score Exceeded (Total Score: 10)"] [data ""] [severity "2"] [ver "OWASP_CRS/3.3.2"] [maturity "0"] [accuracy "0"] [tag "modsecurity"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-generic"] [unique_id "1647007289"] [ref ""] `

**To Reproduce**

`curl "ANY_WORKING_API_WITH_NGINX_MODSECURITY" -w "\n%{time_connect},%{time_total},%{speed_download},%{http_code},%{size_download},%{url_effective}\n" --data-binary @bigJsonFile.json`
(see attachment file [bigJsonFile.txt](https://github.com/SpiderLabs/ModSecurity/files/8233192/bigJsonFile.txt) )

**Expected behavior**

I expect that the request does not hang for ~3 minutes.

**Server :**
 - ModSecurity version (and connector): ModSecurity v3.0.6 with nginx-connector unknown (not found in the image https://hub.docker.com/r/owasp/modsecurity)
 - WebServer: nginx-1.20.2
 - OS (and distro): Linux, Ubuntu with Docker

**Rule Set (please complete the following information):**
 - Running any public or commercial rule set? Public
 - What is the version number? No idea

Thanks you for any help

**Comment by martinhsv** [comment]
@regazzoj ,

If the Content-Type header says 'application/w-www-form-urlencoded', but the request body is, in fact, not that, then ModSecurity will nevertheless attempt to parse the body as if it were in the form parm1=val1&parm2=val&...

You have a few option available to deal with such issues. A couple of the simplest:
- you could use 'SecArgumentsLimit 1000' (or possibly lower value if that is suitable for you). That will only be effective if the body content contains sufficient '&' characters.
- if you are using v3/master (or wish to experiment with it) rather than the tagged version of v3.0.6 from Nov. 2021, SecRequestBodyNoFilesLimit is available there

**Comment by regazzoj** [comment]
@martinhsv ,
Thank you for your answer.

I see where my problem is now, and that's good.

As you suggest, I will try to decrease the arguments limit to see if there is any change. I am just wondering if the fact that it is only a JSON containing an array of objects, so without any '&' to separate values, is it worth to try to change this ?

And if it doesn't work, I will try to create a docker image to use v3/master branch instead of the tagged version 3.0.6.

May I ask if you have any idea when will be available the next release (with SecRequestBodyNoFilesLimit) ?

**Comment by martinhsv** [comment]
The next 3.0.x release is still at least a month away.

In dealing with your delay situation some other options are available. Which ones would be useful partly depends on whether i) the parsing of the request body itself is causing a large delay or ii) running that parsed data against a large-ish rule set is where the delay is.

In your case, it's likely that it's the second thing. If that is indeed the case you could also do something like:
- create a phase:2 rule that runs before your other phase:2 rules, have it do a size check (is REQUEST_BODY greater size than whatever number; check the Content-Length header value; check for an individual ARG longer than some value), and then either deny or disable other phase:2 rules as you choose.

**Comment by regazzoj** [comment]
Okay, that's good to know ! Thanks.

With your explanations, I also think that we are in the second explanation, because we didn't disable any rule.

I will try to do as you said and creating a phase 2 rule to check for the Content-Length header value.

Thanks a lot !

**Comment by regazzoj** [comment]
@martinhsv with your help we manage to fix our issue.

As you suggest, we manage to create a docker image of Modsecurity using 'v3/master' branch.
So our 'SecRequestBodyNoFilesLimit' was working as expected, and so our problem, for this peculiar problem was solved.
But we found that this limit was too low for other APIs, and we decide to just deny access to anyone sending a request with 'x-www-form-urlencoded' as Content-Type header.

And now, it doesn't hang and work like a charm !

Thanks for your time and your help.
I really appreciate !

Have a nice day (:
